libs/agent-sm/agent/src/main/java/org/opensearch/javaagent/SocketChannelInterceptor.java

```diff
@@ -0,0 +1,72 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.javaagent;
+
+import org.opensearch.javaagent.bootstrap.AgentPolicy;
+
+import java.lang.StackWalker.Option;
+import java.lang.reflect.Method;
+import java.net.InetSocketAddress;
+import java.net.NetPermission;
+import java.net.SocketPermission;
+import java.net.UnixDomainSocketAddress;
+import java.security.Policy;
+import java.security.ProtectionDomain;
+import java.util.List;
+
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.asm.Advice.Origin;
+
+/**
+ * SocketChannelInterceptor
+ */
+public class SocketChannelInterceptor {
+    /**
+     * SocketChannelInterceptor
+     */
+    public SocketChannelInterceptor() {}
+
+    /**
+     * Interceptors
+     * @param args arguments
+     * @param method method
+     * @throws Exception exceptions
+     */
+    @Advice.OnMethodEnter
+    @SuppressWarnings("removal")
+    public static void intercept(@Advice.AllArguments Object[] args, @Origin Method method) throws Exception {
+        final Policy policy = AgentPolicy.getPolicy();
+        if (policy == null) {
+            return; /* noop */
+        }
+
+        final StackWalker walker = StackWalker.getInstance(Option.RETAIN_CLASS_REFERENCE);
+        final List<ProtectionDomain> callers = walker.walk(new StackCallerChainExtractor());
```

Comment:
instead of getting all the Protection domains: which may have performance penalty to take deep walks, can we stop walking when we see the first domain with violation?

I was wondering to change the stack walker to something like below: 

```

    public static Optional<ProtectionDomain> findFirstViolation(Policy policy, Permission permission) {
        StackWalker walker = StackWalker.getInstance(StackWalker.Option.RETAIN_CLASS_REFERENCE);
        
        return walker.walk(frames -> 
            frames
                  .map(StackFrame::getDeclaringClass)
                  .map(Class::getProtectionDomain)
                  .filter(pd -> pd != null && !policy.implies(pd, permission))
                  .findFirst()
        );
    }
```